### ISSUE TITLE
polymer template file not found if PolymerTemplate java class is referenced in a spring managed bean

### ISSUE THREAD

**Issue by Klaudeta** [vaadin/flow#6585]
Steps to reproduce:

1. A spring  application using Vaadin 14.x
2. A PolymerTemplate component declared as:

```
@Tag("my-template")
@JsModule("./src/my-template.js")
public class MyTemplate extends PolymerTemplate<MyTemplate.MyTemplateModel> {
......
}
```

3. A managed bean component containing the PolymerTemplate and the interface used to autowire the bean
```
public interface SomeInterface  {
    Component asComponent();
}
```

```
@SpringComponent
@Scope("prototype")
public class SomeManagedBean extends VerticalLayout implements SomeInterface {

    public SomeManagedBean() {
        add(new MyTemplate());
    }

    @Override
    public Component asComponent() {
        return this;
    }
}
```
and finally the Route class:
```
@Route
@PWA(name = "Project Base for Vaadin Flow with Spring", shortName = "Project Base")
public class MainView extends VerticalLayout {

    public MainView(SomeInterface someInterface) {
        this.someInterface = someInterface;
        add(someInterface.asComponent());
    }
}
```

When running the app and navigating to the MainView route, an exception is thrown complaining:
```
Couldn't find the definition of the element with tag 'my-template' in any template file declared using '@JsModule' annotations
```

Minimal project to reproduce the issue
[skeleton-starter-flow-spring-polymertemplate-spring-managed-bean.zip](https://github.com/vaadin/flow/files/3672699/skeleton-starter-flow-spring-polymertemplate-spring-managed-bean.zip)

**Comment by Legioth** [comment]
Duplicate of https://github.com/vaadin/flow/issues/5658.

https://github.com/vaadin/flow/issues/6482, which is tentatively planned for Vaadin 14.1, will cover-up this situation at the expense of requiring the end user to download a larger production bundle.

To get a production bundle that only contains actually used components, you need to use `@Uses` or similar. This way of fixing the problem can be used already today, and will be needed also in the future to avoid downloading redundant component implementations.

**Comment by Legioth** [comment]
Duplicate of https://github.com/vaadin/flow/issues/5658. The problem is that the build doesn't know that `SomeManagedBean` is actually used by the application and thus also doesn't know anything about `MyTemplate`.

https://github.com/vaadin/flow/issues/6482 which is planned for Vaadin 14.1 will reduce the impact in this kind of situation, but at the expense of forcing the end user to download a bundle containing all components from the classpath rather than only the used ones.

You can use `@Uses` to let the build know which component classes are actually used. Right now, that will make the error go away. Once #6482 has been implemented, `@Uses` will only help avoid downloading a bundle that is bigger than necessary.

**Comment by Legioth** [comment]
I will actually leave this issue open since we might want to also improve the error message to state the same things about reachability that is also stated in cases when a "missing" `@JsModule` is encountered but isn't needed for `PolymerTemplate`.

**Comment by klaussemarsse** [comment]
If I downgrade the project to V13 and preserve same exact structure of the Java classes and only change the js template to a polymer 2 html template, everything works fine. So somehow the internal implementation of V13 is able to detect the `html` file although `it doesn't know that that SomeManagedBean is actually used by the application and thus also doesn't know anything about MyTemplate.`
Having said this the problem becomes much an issue when trying to upgrade to V14 a rather complex spring application that uses a lot of managed beans providing components that are not detected as being used during build time.

**Comment by denis-anisimov** [comment]
This is definitely a problem.
I totally agree with you that it's quite confusing and annoying.
It's caused by the byte code scanner which is currently used to find all class usages and it's not able to do it in any DI.

So several notes here:
- V13 uses webjars and that's the reason why it works out of the box: everything is already in the classpath in "binary" format.
- we have already changed the strategy from byte code scanning to full class path scanning in dev mode
- byte code scanning strategy is still used by default for the production mode .
- you may configure which strategy you want to use.
- even with byte code scanner there will be a fallback which will allow to avoid such issue.

This is what we are working right and it will be released soon.
I'm not sure but full classpath strategy for dev mode may be is already released.

**Comment by klaussemarsse** [comment]
I'm not running the app in production mode. 
In which version is the strategy changed in dev mode?
How do I check what strategy I'm using and how to I configure the strategy to avoid the issue?

**Comment by Klaudeta** [comment]
I'm not running the app in production mode.
In which version is the strategy changed in dev mode?
How do I check what strategy I'm using and how to I configure the strategy to avoid the issue?

**Comment by denis-anisimov** [comment]
Before the fix it doesn't matter which mode you are running the app.
It always uses byte code scanner.

With the fix dev mode is using full CP scanning, production mode is using byte code scanning.
It may be changed with the parameter `optimizeBundle`.

I don't follow the releases so I don't know which release has the fix.
Please use the releases page and check which version contains the fix.
https://github.com/vaadin/flow/releases

https://github.com/vaadin/flow-and-components-documentation/blob/master/documentation/production/tutorial-production-mode-advanced.asciidoc#goal-parameters-1  (`optimizeBundle`)
https://github.com/vaadin/flow-and-components-documentation/blob/master/documentation/advanced/tutorial-all-vaadin-properties.asciidoc

**Comment by Legioth** [comment]
The difference between Vaadin 13 and Vaadin 14 is that 13 includes everything from the classpath in the bundle that the end user will download (in production), whereas 14 tries to only include the parts that are actually used (and this is also reflected during development to avoid surprises later on).

This approach works well and gives download size improvements in most cases, but causes this kind of problem in projects that hide the actual component implementations behind interfaces. That's why we're planning to add a new feature to Vaadin 14.1 that makes it automatically fall back to downloading everything from the classpath when necessary.

**Comment by joheriks** [comment]
Note that this feature has not been released yet, it is slated to land in 2.1.0.beta2.

**Comment by denis-anisimov** [comment]
Duplicate of #5658 which is fixed.
